Tools/autotest/arduplane.py

```diff
@@ -1745,6 +1745,16 @@ def fly_soaring(self):
 
         self.progress("Mission OK")
 
+    def fly_terrain_mission(self):
+
+        self.customise_SITL_commandline([], wipe=True)
```

Comment:
Why is this wipe required?

State from previous tests should be reverted as part of the test framework.  This isn't a blocker, but it would be nice to know why so we can fix the test framework.

Reply:
In my initial testing the parameters were not reset from previous tests. I have rebased since then, so I will see if I can roll back this change

Reply:
It still seems that the parameters persist between tests without my added wipe - see this branch https://github.com/samuelctabor/ardupilot/tree/terrain-autotest-ci

In this branch, the terrain test passes individually but fails as part of the full suite. Extracting the parameters from the logs and diffing, I can see that the changes from previous tests are still there. In this case it's SOAR_ENABLE that causes the terrain test to fail.

Looking at [run_one_test](https://github.com/ArduPilot/ardupilot/blob/master/Tools/autotest/common.py#L3863), it seems that the SITL instance is reset only if the previous test customised the arguments, and [reset_SITL_commandline](https://github.com/ArduPilot/ardupilot/blob/master/Tools/autotest/common.py#L1439
)  does not wipe the parameters.

Reply:
OK, so this is a bug in the autotest framework.

I'm working up a PR now.

Problem is mismatched context push vs context pop.  I'm adding a sanity check for that, too.

Reply:
OK, that context was only one of (at least) two issues.

Using a different defaults parameter file (as we do in soaring) when customising the SITL commandline is problematic as we can't tell what's gone into permanent storage and what hasn't.  Further, when we context-push and context-pop, we stash away the value so we can revert it - but if that has come from the replacement defaults file then it isn't what we want in place when we reboot.

Blowing away all parameters as we enter a test isn't really the correct answer - if someone were to insert a test between soaring and this one they'd get the same failure being fixed here.

What we probably need to do is make customising  the SITL command-line more context-like - when you pop up from that context anything done within it is lost (including any parameter-sets).  This will be accomplished by copying away the persistent storage file and moving it back when the pop is done.

In the meantime - could we do that wipe at the *end* of Soaring rather than the start of Terrain?

(Side-note, BTW.  This wouldn't work on Copter, as it applies its default parameters as actual set-parameters rather than specifying a defaultsfile; wiping the parameters would lose you all of your default parameters! :-) )